cirq-google/cirq_google/engine/abstract_local_engine.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import datetime
+from typing import Dict, List, Optional, Sequence, Set, Union, TYPE_CHECKING
+
+import cirq
+from cirq_google.engine.abstract_job import AbstractJob
+from cirq_google.engine.abstract_program import AbstractProgram
+from cirq_google.engine.abstract_local_processor import AbstractLocalProcessor
+from cirq_google.engine.abstract_engine import AbstractEngine
+from cirq_google.engine.client import quantum
+from cirq_google.serialization import Serializer
+
+
+if TYPE_CHECKING:
+    import cirq_google
+    import google.protobuf
+
+
+class AbstractLocalEngine(AbstractEngine):
+    """Collection of processors that can execute quantum jobs.
+
+    This class assumes that all processors are local.  Processors
+    are given during initialization.  Program and job querying
+    functionality is done by serially querying all child processors.
+
+    """
+
+    def __init__(self, processors: List[AbstractLocalProcessor]):
+        for processor in processors:
+            processor.set_engine(self)
+        self._processors = {proc.processor_id: proc for proc in processors}
+
+    def get_program(self, program_id: str) -> AbstractProgram:
```

Comment:
Should `AbstractProgram` here be `AbstractLocalProgram`?

Similarly for other non-local types in this file.

Reply:
I think this would violate the inheritance structure since this is defined in the interface as to return AbstractProgram.  AbstractLocalProgram is a subtype of AbstractProgram, so this is still true.